### DISCUSSION TITLE
Extension not working in release

### DISCUSSION THREAD

**Discussion by roukrak** [microsoft/vscode-discussions#954] [Extension Development QnA] [answered]
I'm writing my first VS Code Extension. When developing and running it (debug) locally it works just fine. The extension adds an icon to the activityBar to open a sidebar with a `TreeItem` "Wallet". Clicking on it opens a `Webview` where users can generate a testnet faucet wallet.

Now I released the project to the marketplace and installed it I can see the icon in the `activityBar`. But now the sidebar does not load any `TreeItems`, just stating " There is no data provider registered that can provide view data. ". When trying to refresh the `TreeDataProvider` via a VS Code command (Ctrl+Shift+P) I only get a popup stating "command xrpl-tools.refresh" not found. This message comes for every command my extension should provide.

package.json
```json
{
  "name": "xrpl-tools",
  "description": "A vscode extension providing tools to develop on the XRP Ledger",
  "version": "1.0.4",
  "engines": {
    "vscode": "^1.84.2"
  },
  "categories": [
    "Other"
  ],
  "main": "./dist/extension.js",
  "publisher": "LimpidCrypto",
  "contributes": {
    "viewsContainers": {
      "activitybar": [
        {
          "id": "xrplToolsContainer",
          "title": "XRPL Tools",
          "icon": "assets/images/logo.png"
        }
      ]
    },
    "views": {
      "xrplToolsContainer": [
        {
          "id": "xrpl-tools.main",
          "name": "xrpl"
        }
      ]
    },
    "commands": [
      {
        "command": "xrpl-tools.refresh",
        "title": "XRPL Tools: Refresh"
      },
      {
        "command": "xrpl-tools.openWalletView",
        "title": "XRPL Tools: Open wallet view"
      },
      {
        "command": "xrpl-tools.generateFaucetWallet",
        "title": "XRPL Tools: Generate faucet wallet"
      }
    ]
  },
  "activationEvents": [
    "onView:xrplToolsContainer"
  ],
  "scripts": {
    "install:all": "npm install && cd webview-ui && npm install",
    "start:webview": "cd webview-ui && npm run start",
    "build:webview": "cd webview-ui && npm run build",
    "vscode:prepublish": "npm run compile",
    "compile": "tsc -p ./",
    "watch": "tsc -watch -p ./",
    "pretest": "npm run compile && npm run lint",
    "lint": "tslint -c tslint.json -p tsconfig.json"
  },
  "devDependencies": {
    "@types/glob": "^8.1.0",
    "@types/minimatch": "^5.1.2",
    "@types/node": "^20.10.1",
    "@types/vscode": "^1.84.2",
    "codelyzer": "^6.0.2",
    "glob": "^10.3.10",
    "minimatch": "^9.0.3",
    "prettier": "^3.1.0",
    "tslint": "^6.1.3",
    "typescript": "^5.2.0",
    "vscode-test": "^1.6.1"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/limpidcrypto/vsc-xrpl-tools.git"
  },
  "dependencies": {
    "xrpl": "^2.13.0"
  }
}
```

extension.ts
```typescript
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { WalletPanel } from "./panels/walletPanel";
import { TreeDataProvider } from './treeDataProvider';
import { generateFaucetWallet } from './app/wallet';

export function activate(context: vscode.ExtensionContext) {
	const treeDataProvider = new TreeDataProvider();
	vscode.window.registerTreeDataProvider('xrpl-tools.main', treeDataProvider);

	let refreshCommand = vscode.commands.registerCommand('xrpl-tools.refresh', () => {
		treeDataProvider.refresh();
	});

	let openWalletCommand = vscode.commands.registerCommand('xrpl-tools.openWalletView', () => {
		new WalletPanel(context.extensionUri);
	});
	let generateFaucetWalletCommand = vscode.commands.registerCommand('xrpl-tools.generateFaucetWallet', async () => {
        const faucetWallet = await generateFaucetWallet();
        vscode.window.showInformationMessage(`Address: ${faucetWallet.wallet.classicAddress}\nSecret: ${faucetWallet.wallet.seed ?? faucetWallet.wallet.privateKey}\nFunded with: ${faucetWallet.balance}`);
	});

	context.subscriptions.push(refreshCommand);
	context.subscriptions.push(openWalletCommand);
	context.subscriptions.push(generateFaucetWalletCommand);
}

export function deactivate() { }
```

treeDataProvider.ts
```typescript
import * as vscode from 'vscode';

export class TreeDataProvider implements vscode.TreeDataProvider<vscode.TreeItem> {
    private _onDidChangeTreeData: vscode.EventEmitter<vscode.TreeItem | undefined> = new vscode.EventEmitter<vscode.TreeItem | undefined>();
    readonly onDidChangeTreeData: vscode.Event<vscode.TreeItem | undefined> = this._onDidChangeTreeData.event;

    refresh(): void {
        this._onDidChangeTreeData.fire(undefined);
    }

    getTreeItem(element: vscode.TreeItem): vscode.TreeItem {
        if (element.label === 'Wallet') {
            element.command = {
                command: 'xrpl-tools.openWalletView',
                title: 'XRPL Tools: Open Wallet',
                arguments: [element]
            };
        }
        return element;
    }

    getChildren(element?: vscode.TreeItem): vscode.TreeItem[] {
        const walletTreeItem = new vscode.TreeItem('Wallet');
        walletTreeItem.iconPath = new vscode.ThemeIcon('account');
        return [walletTreeItem];
    }
}
```

Anyone ran into a similar problem once? What other files/info do you need?

Thank you!

**Comment by alefragnani** [comment, marked as answer]
Hi @roukrak,

Once you're receiving a `command xrpl-tools.refresh` error message, it suggests your extension couldn't be properly activated. I would suggest you to start looking for erros in the `Console` window, opened via `Developer: Toggle Developer Tools` in the VS Code instance that has the extension installed. 

And, looking at your `package.json` file, I see a dependency to `xrpl`. I don't know this package, but you may know that it is necessary for you to be sure any additional dependency to be _packaged_ with the extension. Because you are not using any bundler (WebPack / esbuild), and by default the `node_modules` is not _packaged_ while running `vsce package` / `vsce publish`, maybe this dependency is not available in the released extension. So, open the extension's folder with `Extensions: Open Extensions Folder`, and take a look if it is there. Or simply run `vsce ls` in your extension's root folder and you will see which files are being _packaged_.

Hope this helps

**Reply by roukrak** [reply]
That was exactly my issue. Thanks for your help :)

**Comment by Neobank21** [comment]
> I'm writing my first VS Code Extension. When developing and running it (debug) locally it works just fine. The extension adds an icon to the activityBar to open a sidebar with a `TreeItem` "Wallet". Clicking on it opens a `Webview` where users can generate a testnet faucet wallet.
> 
> 
> 
> Now I released the project to the marketplace and installed it I can see the icon in the `activityBar`. But now the sidebar does not load any `TreeItems`, just stating " There is no data provider registered that can provide view data. ". When trying to refresh the `TreeDataProvider` via a VS Code command (Ctrl+Shift+P) I only get a popup stating "command xrpl-tools.refresh" not found. This message comes for every command my extension should provide.
> 
> 
> 
> package.json
> 
> ```json
> 
> {
> 
>   "name": "xrpl-tools",
> 
>   "description": "A vscode extension providing tools to develop on the XRP Ledger",
> 
>   "version": "1.0.4",
> 
>   "engines": {
> 
>     "vscode": "^1.84.2"
> 
>   },
> 
>   "categories": [
> 
>     "Other"
> 
>   ],
> 
>   "main": "./dist/extension.js",
> 
>   "publisher": "LimpidCrypto",
> 
>   "contributes": {
> 
>     "viewsContainers": {
> 
>       "activitybar": [
> 
>         {
> 
>           "id": "xrplToolsContainer",
> 
>           "title": "XRPL Tools",
> 
>           "icon": "assets/images/logo.png"
> 
>         }
> 
>       ]
> 
>     },
> 
>     "views": {
> 
>       "xrplToolsContainer": [
> 
>         {
> 
>           "id": "xrpl-tools.main",
> 
>           "name": "xrpl"
> 
>         }
> 
>       ]
> 
>     },
> 
>     "commands": [
> 
>       {
> 
>         "command": "xrpl-tools.refresh",
> 
>         "title": "XRPL Tools: Refresh"
> 
>       },
> 
>       {
> 
>         "command": "xrpl-tools.openWalletView",
> 
>         "title": "XRPL Tools: Open wallet view"
> 
>       },
> 
>       {
> 
>         "command": "xrpl-tools.generateFaucetWallet",
> 
>         "title": "XRPL Tools: Generate faucet wallet"
> 
>       }
> 
>     ]
> 
>   },
> 
>   "activationEvents": [
> 
>     "onView:xrplToolsContainer"
> 
>   ],
> 
>   "scripts": {
> 
>     "install:all": "npm install && cd webview-ui && npm install",
> 
>     "start:webview": "cd webview-ui && npm run start",
> 
>     "build:webview": "cd webview-ui && npm run build",
> 
>     "vscode:prepublish": "npm run compile",
> 
>     "compile": "tsc -p ./",
> 
>     "watch": "tsc -watch -p ./",
> 
>     "pretest": "npm run compile && npm run lint",
> 
>     "lint": "tslint -c tslint.json -p tsconfig.json"
> 
>   },
> 
>   "devDependencies": {
> 
>     "@types/glob": "^8.1.0",
> 
>     "@types/minimatch": "^5.1.2",
> 
>     "@types/node": "^20.10.1",
> 
>     "@types/vscode": "^1.84.2",
> 
>     "codelyzer": "^6.0.2",
> 
>     "glob": "^10.3.10",
> 
>     "minimatch": "^9.0.3",
> 
>     "prettier": "^3.1.0",
> 
>     "tslint": "^6.1.3",
> 
>     "typescript": "^5.2.0",
> 
>     "vscode-test": "^1.6.1"
> 
>   },
> 
>   "repository": {
> 
>     "type": "git",
> 
>     "url": "https://github.com/limpidcrypto/vsc-xrpl-tools.git"
> 
>   },
> 
>   "dependencies": {
> 
>     "xrpl": "^2.13.0"
> 
>   }
> 
> }
> 
> ```
> 
> 
> 
> extension.ts
> 
> ```typescript
> 
> // The module 'vscode' contains the VS Code extensibility API
> 
> // Import the module and reference it with the alias vscode in your code below
> 
> import * as vscode from 'vscode';
> 
> import { WalletPanel } from "./panels/walletPanel";
> 
> import { TreeDataProvider } from './treeDataProvider';
> 
> import { generateFaucetWallet } from './app/wallet';
> 
> 
> 
> export function activate(context: vscode.ExtensionContext) {
> 
> 	const treeDataProvider = new TreeDataProvider();
> 
> 	vscode.window.registerTreeDataProvider('xrpl-tools.main', treeDataProvider);
> 
> 
> 
> 	let refreshCommand = vscode.commands.registerCommand('xrpl-tools.refresh', () => {
> 
> 		treeDataProvider.refresh();
> 
> 	});
> 
> 
> 
> 	let openWalletCommand = vscode.commands.registerCommand('xrpl-tools.openWalletView', () => {
> 
> 		new WalletPanel(context.extensionUri);
> 
> 	});
> 
> 	let generateFaucetWalletCommand = vscode.commands.registerCommand('xrpl-tools.generateFaucetWallet', async () => {
> 
>         const faucetWallet = await generateFaucetWallet();
> 
>         vscode.window.showInformationMessage(`Address: ${faucetWallet.wallet.classicAddress}\nSecret: ${faucetWallet.wallet.seed ?? faucetWallet.wallet.privateKey}\nFunded with: ${faucetWallet.balance}`);
> 
> 	});
> 
> 
> 
> 	context.subscriptions.push(refreshCommand);
> 
> 	context.subscriptions.push(openWalletCommand);
> 
> 	context.subscriptions.push(generateFaucetWalletCommand);
> 
> }
> 
> 
> 
> export function deactivate() { }
> 
> ```
> 
> 
> 
> treeDataProvider.ts
> 
> ```typescript
> 
> import * as vscode from 'vscode';
> 
> 
> 
> export class TreeDataProvider implements vscode.TreeDataProvider<vscode.TreeItem> {
> 
>     private _onDidChangeTreeData: vscode.EventEmitter<vscode.TreeItem | undefined> = new vscode.EventEmitter<vscode.TreeItem | undefined>();
> 
>     readonly onDidChangeTreeData: vscode.Event<vscode.TreeItem | undefined> = this._onDidChangeTreeData.event;
> 
> 
> 
>     refresh(): void {
> 
>         this._onDidChangeTreeData.fire(undefined);
> 
>     }
> 
> 
> 
>     getTreeItem(element: vscode.TreeItem): vscode.TreeItem {
> 
>         if (element.label === 'Wallet') {
> 
>             element.command = {
> 
>                 command: 'xrpl-tools.openWalletView',
> 
>                 title: 'XRPL Tools: Open Wallet',
> 
>                 arguments: [element]
> 
>             };
> 
>         }
> 
>         return element;
> 
>     }
> 
> 
> 
>     getChildren(element?: vscode.TreeItem): vscode.TreeItem[] {
> 
>         const walletTreeItem = new vscode.TreeItem('Wallet');
> 
>         walletTreeItem.iconPath = new vscode.ThemeIcon('account');
> 
>         return [walletTreeItem];
> 
>     }
> 
> }
> 
> ```
> 
> 
> 
> Anyone ran into a similar problem once? What other files/info do you need?
> 
> 
> 
> Thank you!
